python/tvm/contrib/cc.py

```diff
@@ -51,9 +51,32 @@ def create_shared(output,
     else:
         raise ValueError("Unsupported platform")
 
+def get_target_triple():
+    """ Get the target triple using compiler.
+
+    Returns
+    -------
+        out: str (Linux / Mac) or None (Win32)
+    """
+    if sys.platform == "darwin" or sys.platform.startswith("linux"):
+        cmd = ["g++", "-dumpmachine"]
+        proc = subprocess.Popen(
+            cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+        (out, _) = proc.communicate()
+        if proc.returncode != 0:
+            msg = "dumpmachine error:\n"
+            msg += py_str(out)
+            raise RuntimeError(msg)
+        return py_str(out)
+    elif sys.platform == "win32":
+        return None
+    else:
+        raise ValueError("Unsupported platform")
+
 
 # assign so as default output format
 create_shared.output_format = "so" if sys.platform != "win32" else "dll"
+create_shared.get_target_triple = get_target_triple()
```

Comment:
Let us do it lazily(otherwise user who import cc implicitly and does not have a proper compiler tool chain will get an error)

create_shared.get_target_triple = get_target_triple

Reply:
Let us create a functor that returns this function and reuse it in most places
```
def get_target_by_dump_machine(compiler):
     def get_target_triple():
         ...
     return get_target_triple
```

Reply:
please also add support for cross_compiler in this file

Reply:
> please also add support for cross_compiler in this file

In my mind, I think get_target_triple should be used for host running TVM or we call ndk (we could know the compiler according to TVM_NDK_CC). Like we use cross_compiler (currently be used by MicroDev) to generate C file, we don't know the compiler. If I don't understand correctly of cross_compiler's effect, please correct me. Thanks!

Reply:
I specificlaly mean add support here https://github.com/apache/incubator-tvm/blob/master/python/tvm/contrib/cc.py#L107

Reply:
> https://github.com/apache/incubator-tvm/blob/master/python/tvm/contrib/cc.py#L107

Ok. combine your context I think you mean:

```python
def get_target_by_dump_machine(compiler="g++"): # default value be g++. could be specified by users
    def get_target_triple():
    if sys.platform == "darwin" or sys.platform.startswith("linux"):
        cmd = [compiler, "-dumpmachine"] # change to compiler
        proc = subprocess.Popen(
            cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
        (out, _) = proc.communicate()
        if proc.returncode != 0:
            msg = "dumpmachine error:\n"
            msg += py_str(out)
            raise RuntimeError(msg)
        return py_str(out)
    elif sys.platform == "win32":
        return None
    else:
        raise ValueError("Unsupported platform")
return get_target_triple
# lazily create
create_shared.get_target_by_dump_machine = get_target_by_dump_machine
def cross_compiler():
     ...
    fcompile.get_target_by_dump_machine = get_target_by_dump_machine
```

Also we need to add this support for NDK and `build_create_shared_func` so that all fcompile could have this fn support.

Reply:
> Sorry I specificlaly mean add support here https://github.com/apache/incubator-tvm/blob/master/python/tvm/contrib/cc.py#L107
> 
> By allowing passing in an optional get_target_triple function

Sorry I just saw your recent change of `By allowing passing in an optional get_target_triple function`. I understood previous that we should only expose `get_target_by_dump_machine` and `get_target_triple` should only accept `compiler` parameter. Seems that you want we could accept one functor of `get_target_triple` in cross_compiler so that we could get this information, i.e.

```python
def get_target_by_dump_machine(compiler="g++"): # default value be g++. could be specified by other api
    def get_target_triple():
    if sys.platform == "darwin" or sys.platform.startswith("linux"):
        cmd = [compiler, "-dumpmachine"] # change to compiler
        proc = subprocess.Popen(
            cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
        (out, _) = proc.communicate()
        if proc.returncode != 0:
            msg = "dumpmachine error:\n"
            msg += py_str(out)
            raise RuntimeError(msg)
        return py_str(out)
    elif sys.platform == "win32":
        return None
    else:
        raise ValueError("Unsupported platform")

def cross_compiler(compile_func, base_options=None, output_format="so",  get_target_triple=None):
 ...
_fcompile.get_target_by_dump_machine = get_target_triple if get_target_triple else get_target_by_dump_machine
```

Reply:
```python
create_shared.get_target_triple = get_target_by_dump_machine(
     "g++" if sys.platform == "darwin" or sys.platform.startswith("linux") else None)

ndk.get_target_triple = get_target_by_dump_machine(NDK_CC)

```

Reply:
In the case of _fcompile, we can simply assign it to get_target_triple if user pass it in, or assign it to None

Reply:
Understand your meaning. Thanks. Will create new pr after I finish lunch.